src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Value/VersionConverter.cs

```diff
@@ -1,16 +1,78 @@
+
+#if BUILDING_INBOX_LIBRARY
+        private const int MaxStringLengthOfPositiveInt32 = 10; // int.MaxValue.ToString().Length
+
+        private const int
+            MaxStringLengthOfVersion = (MaxStringLengthOfPositiveInt32 * VersionComponentsCount) + 1 + 1 + 1; // 43, 1 is length of '.'
+#endif
+
         public override Version Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
         {
-            string? versionString = reader.GetString();
-            if (Version.TryParse(versionString, out Version? result))
+            ReadOnlySpan<byte> source = reader.HasValueSequence ? reader.ValueSequence.ToArray() : reader.ValueSpan;
+
+            Span<int?> versionComponents = stackalloc int?[VersionComponentsCount] { null, null, null, null };
+            int indexOfDot = GetIndexOfDot(source);
+
+            if (reader._stringHasEscaping || indexOfDot == -1)
             {
-                return result;
+                ThrowHelper.ThrowJsonException();
+            }
+
+            Debug.Assert(source.IndexOf(JsonConstants.BackSlash) == -1, "Version value should have no escaping");
+            Debug.Assert(indexOfDot != -1, "Version should have at least Major and Minor values separated by \".\"");
+
+            for (int i = 0; i < versionComponents.Length; i++)
+            {
+                bool lastComponent = indexOfDot == -1;
+                var readOnlySpan = lastComponent ? source : source.Slice(0, indexOfDot);
+                if (Utf8Parser.TryParse(readOnlySpan, out int value, out _))
+                {
+                    if (value < 0)
+                    {
+                        ThrowHelper.ThrowJsonException();
+                    }
+
+                    versionComponents[i] = value;
+                    source = source.Slice(indexOfDot + 1);
+                    indexOfDot = GetIndexOfDot(source);
+                    if (lastComponent)
+                    {
+                        break;
+                    }
+                }
+                else
+                {
+                    ThrowHelper.ThrowJsonException();
+                }
+            }
+
+            ref int? major = ref versionComponents[0];
+            ref int? minor = ref versionComponents[1];
+            ref int? build = ref versionComponents[2];
+            ref int? revision = ref versionComponents[3];
+            if (major.HasValue && minor.HasValue && build.HasValue && revision.HasValue)
+            {
+                return new Version(major.Value, minor.Value, build.Value, revision.Value);
+            }
+            else if (major.HasValue && minor.HasValue && build.HasValue)
+            {
+                return new Version(major.Value, minor.Value, build.Value);
+            }
+            else if (major.HasValue && minor.HasValue)
+            {
+                return new Version(major.Value, minor.Value);
```

Comment:
I'm not understanding why all this is necessary.  Can you clarify?  Why can't we use Version.TryParse's span-based overload?

Reply:
Yes, you are right, we can encode bytes to chars and pass them to Version.TryParse, for some reason i haven't thought of it. I'll make changes and push them here.

Reply:
Fixed in 07b6a9a